SEAICE_USE_GROWTH_ADX: fix and improve diagnostics for mass and heat budget

antnguyen13 opened the pull request:
## What changes does this PR introduce?
Fix bug associated with incorrect inclusion of sublimation term in diagnostics of sea ice mass budget and changed name of diagnostic term for sea ice heat budget associated with EmPmR in seaice_growth_adx.F. 

## What is the current behaviour? 
1. The offline mass budget for sea ice is NOT closed if SEAICE_USE_GROWTH_ADX is defined in SEAICE_OPTIONS.h.  This is due to the inclusion of the a term related to sublimation, whose physics were NOT accounted for in seaice_growth_adx.F, into the diagnostic SIatmFW.  

2. The heat flux term associated with EmPmR in seaice_growth_adx.F is being diagnosed to SIaaflux, with the latter being used to represent omitted physics related to sublimation in seaice_growth.F. This causes confusion when trying to close the heat budget in the latter case as the user might make mistake thinking the term is related to sublimation and therefore needing to understand why the term is needed to close the heat budget.

## What is the new behaviour 
1. The sublimation contribution to the diagnostic SIatmFW (required to close offline sea ice mass budget) is removed.

2. A new variable diagnostic SIeprflx is defined in seaice_diagnostics_init.F and filled in seaice_growth_adx.F for heat associated with EmPmR, and the local variable SIaaflux is completely removed from seaice_growth_adx.F.

## Does this PR introduce a breaking change? 
Not that I am aware of for online.  This PR primarily affects diagnostic terms , e.g., the offline post-processing to close mass and heat budgets for sea ice.

## Other information:
Both ASTE and ECCOv4r* have been using seaice_growth_adx in their respective local code with the sublimation term already commented out (this PR).  However, the bug and confusion definition of SIaaflux have remained in the main MITgcm branch. It's time to fix them.

## Suggested addition to `tag-index`
(To avoid unnecessary merge conflicts, please don't update `tag-index`. One of the admins will do that when merging your pull request.)

Fix diagnostic SIatmFW and add diagnostic SIeprflx to close offline sea ice mass and heat budgets.

antnguyen13 commented:
Ok, i removed all comments except to add a single line just above the actual SEAICE_USE_GROWTH_ADX flag.

mjlosch commented on pkg/seaice/seaice_growth_adx.F:
```diff
@@ -182,9 +182,18 @@ SUBROUTINE SEAICE_GROWTH_ADX( myTime, myIter, myThid )
+catn  In seaice_diagnostics_init.F, SIaaflux is defined only
+catn  if ndef SEAICE_DISABLE_HEATCONSFIX
+catn  Now we define our own SIacflux to avoid confusion with
+catn  mainbranch meaning of SIaaflux.
```
here and elsewhere: it is not clear what "main branch" refers to. Do you mean "seaice_growth.F" as opposed to this routine? If so, please rephrase.

antnguyen13 replied:
ah yes, that's what i meant, will rephrase as in seaice_growth.F . Although I think i'll also do what you suggested, which is removing reference to SIaaflux altogether from this subroutine.

mjlosch commented on pkg/seaice/seaice_growth_adx.F:
```diff
@@ -182,9 +182,18 @@ SUBROUTINE SEAICE_GROWTH_ADX( myTime, myIter, myThid )
       _RL EnergyInNewTotalIceVolume     (1:sNx,1:sNy)
       _RL NetEnergyFluxOutOfOcean       (1:sNx,1:sNy)
 
+catn: these are defined locally here, not global
```
In general, I don't see the point of personalised comments in the code (unless they are not meant to be permanent or a reminder that this will be removed). With "git blame" we can always find out, who added what in the code, in case we need to. (and I know I have added a few "CML" myself)

antnguyen13 replied:
Yes, we have talked before, i'll clean up.  In general i leave the first commit with comments I originally wrote to myself , but after a clean up these should be gone.

mjlosch commented on pkg/seaice/seaice_growth_adx.F:
```diff
@@ -182,9 +182,18 @@ SUBROUTINE SEAICE_GROWTH_ADX( myTime, myIter, myThid )
 CAB   Terms for closing the budget
       _RL SItflux     (1-OLx:sNx+OLx,1-OLy:sNy+OLy,nSx,nSy)
-      _RL SIaaflux    (1-OLx:sNx+OLx,1-OLy:sNy+OLy,nSx,nSy)
+catn Comment out the local definition of SIaaflux as its meaning
```
As the name `SIaaflux` does not seem to be inspired by any meaningful abbreviation, I would just remove any references to this variable in this routine, i.e. just remove and replace by `SIacflux`. (Does `ac` stand for anything?)

antnguyen13 replied:
yes, i think you suggested later as well a more meaningful name, which i'll adopt: change SIacflux to SIeprflx perhaps, to stand for heat term associated with empmr

mjlosch commented on pkg/seaice/seaice_growth_adx.F:
```diff
@@ -1240,18 +1255,32 @@ SUBROUTINE SEAICE_GROWTH_ADX( myTime, myIter, myThid )
      &         * recip_deltaTtherm * HEFFM(I,J,bi,bj)
 
-          SIaaflux(I,J,bi,bj) =  -EmPmR(I,J,bi,bj)
+catn In the main branch, SIaaflux is perserved for ifdef block
```
This comment is a little confusing, because it describes something that does no longer happen here (`SIaaflux` is not defined nor used anymore). I would remove it or say something shorter like: "The corresponding diagnostics in seaice_growth.F contain additional terms that are not accounted for in this routine" (or similar)

mjlosch commented on pkg/seaice/seaice_growth_adx.F:
```diff
@@ -1261,7 +1290,16 @@ SUBROUTINE SEAICE_GROWTH_ADX( myTime, myIter, myThid )
 #endif /* ALLOW_RUNOFF */
      &           )*rhoConstFresh
-     &     +    FWsublim(i,j)* SEAICE_rhoIce * recip_deltaTtherm
+catn The line below is only valid if we account for sublimation in hsnow
```
again, I think you can remove this entire block. No need to compare with seaice_growth.F.
To me it would make much more sense to add a comment at the declaration of `FWsublim`, etc. saying that these variables are required as arguments by `seaice_solve4temp`, but are not used afterwards in this routine (by choice).

antnguyen13 replied:
Yes, thanks, i'll do what you suggest: say FWsublim is required as input for solve4temp but won't be used.  I think the extensive comments i left in my first commit was reflecting my confusion as to why this term was computed and then "used" and yet caused budget issues.. and so comments referring to comparison with seaice_growth.F originally was my effort to trace what was done there and why some code was here because it was copied over but shouldn't be due to missing physics.

mjlosch reviewed (commented):
Thanks for fixing this.
I have some suggestions, the main being that `SIaaflux` need not appear at all in `seaice_growth_adx` anymore, not even in comments.
Maybe `SIacflux`, both variable and diagnostics, can be given a more meaningful name, e.g.,`SIfwhflx` (for freshwater heat flux, or similar).

mjlosch commented:
@antnguyen13 maybe we can (ab-)use this PR to update the documentation with a short section about seaice_growth_adx.F? What do you think?

antnguyen13 commented:
@mjlosch yes, by the weekend, i'll try to add some text to the documentation on this adx code. 
@jm-c & @mjlosch:  A bit related to this PR, there is one more outstanding bug fix and associated diagnostics needed to close the sea ice heat budget, but it's independent of "sublimation" term which is for closing the mass budget (this PR). i think i should wait for this PR to go through before submitting that ?

mjlosch commented:
> A bit related to this PR, there is one more outstanding bug fix and associated diagnostics needed to close the sea ice heat budget, but it's independent of "sublimation" term which is for closing the mass budget (this PR). i think i should wait for this PR to go through before submitting that ?

My suggestion: if it is not independent of this PR, then wait, but if it is independent, then do what is easier for you.

mjlosch reviewed (commented):
Thanks for adding this text. I made a few suggestions and added some tags.

antnguyen13 commented on doc/phys_pkgs/seaice.rst:
```diff
@@ -408,6 +414,18 @@ MITgcm. Heat, fresh water fluxes and surface stresses are computed from the
 atmospheric state and, by default, modified by the ice model at every time
 step.
 
+Within :filelink:`pkg/seaice`, two ice thermodynamic code exists,
```
@mjlosch Could you clarify what this refers to?  

(There is a loose sentence which I copied from the next paragraph that starts with 
"Within :filelink:`pkg/seaice`, two ice thermodynamic code exists,"

and then I realized it was better to grab the sentence from the next subsection describing thsice , so I've now delete this sentence "Within..." from the doc.)

Do you mean to break the thermodynamics and dynamics within pkg/seaice into two subsections?

mjlosch replied:
Much simpler: My suggestion is to move the entire new paragraph down to just before section ``8.6.2.4.1. Compatibility with ice-thermodynamics package pkg/thsice``, because 
1. the order in this is usually first dynamics, then thermodynamics (but that's not important)
2. it is closer to the next section (8.6.2.4.1) , which is about thermodynamics.

antnguyen13 commented on doc/phys_pkgs/seaice.rst:
```diff
@@ -408,6 +414,18 @@ MITgcm. Heat, fresh water fluxes and surface stresses are computed from the
+version seaice_growth_adx.F based on Fenty (2013) :cite:`fenty:13`
+that excludes physics such as ITD, treatment for sublimation, and frazil
+ice but provides a stable sea ice adjointable with physical sensitivity.
+When seaice_growth_adx code is enabled, the regularization parameter
```
thanks. now should be incorporated into next commit (after some clarification request in the next comment from me to you)

antnguyen13 commented on doc/phys_pkgs/seaice.rst:
```diff
@@ -408,6 +414,18 @@ MITgcm. Heat, fresh water fluxes and surface stresses are computed from the
+The :filelink:`pkg/seaice` includes the original so-called zero-layer
+thermodynamics with a snow cover as in the appendix of Semtner (1976)
+:cite:`semtner:76`.  Two versions of this zero-layer thermodynamic code
+exist, with a more developed version seaice_growth.F and a simplified
+version seaice_growth_adx.F based on Fenty (2013) :cite:`fenty:13`
```
Thanks. now modified, soon to be committed.

antnguyen13 commented on doc/phys_pkgs/seaice.rst:
```diff
@@ -408,6 +414,18 @@ MITgcm. Heat, fresh water fluxes and surface stresses are computed from the
 atmospheric state and, by default, modified by the ice model at every time
 step.
 
+Within :filelink:`pkg/seaice`, two ice thermodynamic code exists,
+The :filelink:`pkg/seaice` includes the original so-called zero-layer
+thermodynamics with a snow cover as in the appendix of Semtner (1976)
+:cite:`semtner:76`.  Two versions of this zero-layer thermodynamic code
```
mod made. soon to be committed.

antnguyen13 commented:
@mjlosch thanks, i adapted all your suggestions into the doc.

mjlosch commented:
While we are at it, can we promote the "Dynamics" and "Thermodynamics" parts of the documentation to "Sections" that show up in the table of content, i.e. replace `------` by `======` in the rst-description. It has always bothered me, that you see the TOC as:

```
Introduction
Configuration and compling
Runtime parameters
Description (huge part of the documentation, including details of dynamics and thermodynamics)
Known problems
...
```

which places a wrong "emphasis" on "known problems" (which is actually a very short section). @jrscott, @jm-c?

mjlosch reviewed (approved):
LGTM

antnguyen13 commented:
@jrscott : your move of text is good! i was already struggling to find out where to insert these flags.. the two I put in, freedrift and SInegFac , were essential for running with the sea ice adjoint, that's why they were the only two out of many other flags not there that got added.

@mjlosch thanks for fixing the missing / wrong link! 

On my side, everything is now good to go.  

(As soon as this gets merged i'll proceed with the next PR to fix the last outstanding bug in this adx code (to treat flooding and close heat budget correctly) as well as adapting to linear free surface and virtual salt flux.)

mjlosch commented:
> What do you think of moving Description in front of Configuration and Compiling (as we did for GMREDI doc)?

@jrscott we can try that. If you think that this makes sense (in terms of the flow), I am all for it.

antnguyen13 commented:
I'm going to create another PR on this sea ice and adx , we could wait until then to move sections around in the documentation?  This way i can update my upstream master to include this for my new PR? (otherwise, i 'm worried of another mess i might create if i pre-make my new PR without new upstream master :-) ).

jm-c reviewed (commented):
This looks good. 

I have few minor comments/suggestions, mostly to clarify even further what is specific to `seaice_growth_adx.F`

1.  I would suggest to rename (again) this PR title to:
"SEAICE_USE_GROWTH_ADX: fix and improve diagnostics for mass and heat budget"
This would be a bit shorter and clearly related to this part of the code whereas "seaice ADX" 
might be view as any Adjoint part of pkg/seaice.

2.  The new pkg/seaice diagnostics if not relevant when `SEAICE_USE_GROWTH_ADX` is "#undef".
 I suggest to add the corresponding "#ifdef" in `seaice_diagnostics_init.F`
 so that a user that compiles without SEAICE_USE_GROWTH_ADX defined will not be tempted to turn on this diagnostics that is never filled.
Not sure how to convey this point in the documentation table of diags.

3) I left minor comment below.

jm-c commented:
@antnguyen13  I have a very minor addition to `seaice_readparams.F` (left from PR #666, just to add an "#ifdef ALLOW_COST" to avoid un-used var warnings) that I could push here. It's clearly not related to these changes but not very far (in pkg/seaice, with or without pkg/cost).
But if you prefer not to push this here I would totally understand, with no problem at all.

antnguyen13 commented on doc/phys_pkgs/seaice.rst:
```diff
@@ -66,6 +66,7 @@ options see :filelink:`SEAICE_OPTIONS.h <pkg/seaice/SEAICE_OPTIONS.h>`.
    :varlink:`SEAICE_ALLOW_MCE`, #undef, enable use of Mohr-Coulomb yield curve with elliptical plastic potential
    :varlink:`SEAICE_ALLOW_TD`, #undef, enable use of teardrop and parabolic Lens yield curves with normal flow rules
    :varlink:`SEAICE_LSR_ZEBRA`, #undef, use a coloring method for LSR solver
+   :varlink:`SEAICE_ALLOW_FREEDRIFT`, #undef, solve approximate sea ice momentum equation and bypass solving for sea ice internal stress
```
thanks @jm-c , i added "enable"

antnguyen13 commented:
@jm-c Thanks for the suggestions! I modified the PR title, and added/removed the wording "enable" as you suggested.  I also put a new ifdef SEAICE_USE_GROWTH_ADX around the new diagnostics SIeprflx, *as well as* including an "else" so that SIaaflux is *not* valid , so as to also let the users know that diagnostics is not going to be valid / filled.

I know the original diagnostics SIaaflux was not in the documentation.. and was not sure if i should add the new one in there.. Please let me know if you think I should try to add it?

Also, i'm ok with adding the left-over from PR #666 ifdef to this, i think the commit will say that so that the users wont be confused that it's part of this PR?

Thanks!

antnguyen13 commented on doc/phys_pkgs/seaice.rst:
```diff
@@ -153,6 +157,8 @@ General flags and parameters
   +------------------------------------+------------------------------+-------------------------------------------------------------------------+
   | :varlink:`SEAICEuseFlooding`       | TRUE                         | use flood-freeze algorithm                                              |
   +------------------------------------+------------------------------+-------------------------------------------------------------------------+
+  | :varlink:`SINegFac`                | 1.0                          | over/undershoot factor for sea ice advective term in forward/adjoint    |
```
oh, i overlooked this... let me check how to address...

jm-c commented:
@antnguyen13  From my side, this is good and ready to be merged. If you agree, I could merge this PR relatively soon (like tomorrow ?). Thanks.

antnguyen13 commented:
@jm-c all good on my side! many thanks! (another adx one coming very soon, as soon as this one is merged! but the new one fixes a serious bug! and will come with its own verification other!)
